### PR TITLE
Robot Problem + Test Files

### LINKED ISSUE #24: Robot Problem

### LINKED ISSUE #25: Robot Problem [Creating Files]

### PR DESCRIPTION
**Pull request by reunicorn1**
---
Adding robot.py file to the solutions directory with its corresponding test
---

## Behavior
There is a robot on an m x n grid. The robot is initially located at the top-left corner (i.e., grid[0][0]). The robot tries to move to the bottom-right corner (i.e., grid[m - 1][n - 1]). The robot can only move either down or right at any point in time.

Given the two integers m and n, return the number of possible unique paths that the robot can take to reach the bottom-right corner.

### Files

**solutions/robot.py**

- [O] The file name describes the function's behavior
- [O] There is a module docstring in the function file
- [O] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [O] There is a module docstring in the tests file

### Unit Tests

**solutions/tests/test_robot.py**

- [O] The test class has a helpful name in PascalCase
- [O] The test class has a docstring

### CODE REVIEW

**Comment by reunicorn1** [comment]
**robot.py file**, failed the py format test, as the test marks the second parameter of the function as unused. This might pass at this stage as the function itself isn't written yet. However, I may need better naming for the parameters. Currently I'm using m and n as the problem description states, but I will be using cols and rows to state that I'm referring to the columns and rows of the grid.

**Comment by reunicorn1** [comment]
There is a robot on an m x n grid. The robot is initially located at the top-left corner (i.e., grid[0][0]). The robot tries to move to the bottom-right corner **(i.e., grid[m - 1][n - 1]).** The robot can only move either down or right at any point in time.Given the two integers m and n, return the number of possible unique paths that the robot can take to reach the bottom-right corner.The test cases are generated so that the answer will be less than or equal to 2 * 109.Parameters:

m: The number of rows in the grid.

n: The number of columns in the grid.

Return: The number of unique paths from the top-left corner to the bottom-right corner.

IMG_3919


To solve this problem I used dynamic programming as an approach. The trick is to divide the problem into smaller problems and then reuse your solutions to solve the bigger ones. I used a helper function and recursion to keep track of the movement process. The answer depends on the basis that there's either one road to take (if x == n -1 or y == m - 1 ) or two roads to take otherwise, you save the answers in a memoization dictionary using the grid current location tuple as a key, and utilize these answers without the need to repeat your steps when necessary.

Test cases created also tried to cover the general cases that might be used with the function. Since memoization is used the efficiency of retrieving answers is quite high with an optimum time complexity.

File: solutions/tests/test_robot.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    def test_large_grid(self):
+        """A test for a moderately large grid 10x10"""
+        self.assertEqual(robot(10, 10), 48620)
+
+    def test_very_large_grid(self):
+        """A test for a very large grid 20x15"""
+        self.assertEqual(robot(20, 15), 818809200)
+
+    # Defensive Cases
+    def test_invalid_row(self):
```
Comment (Safi222) [inline solutions/tests/test_robot.py:61]: Great job on handling edge cases

File: solutions/robot.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    # assert the values of rows and clos and make sure they are ints and
+    # greater than 1
+    assert isinstance(rows, int) and rows > 0, "Rows variables has to \
+            be an integer greater than 0"
+    assert isinstance(cols, int) and cols > 0, "Cols variable has to be \
+            greater than 0"
+
+    # use the helper function robot_move to determine the number of paths
+    memo = {}
+    return robot_move((0, 0), rows, cols, memo)  # (0, 0) is the initial spot
```
Comment (Safi222) [inline solutions/robot.py:84]: Good use of a helper function!

File: solutions/tests/test_robot.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+"""
+
+import unittest
+from ..robot import robot
+
+
+class TestRobot(unittest.TestCase):
+    """Test the robot function"""
+
+    # Standard Cases
```
Comment (Safi222) [inline solutions/tests/test_robot.py:22]: Basic case for Single-Cell Grid 1x1 could be added

File: solutions/robot.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        return memo.get(cell)
+    # if the cell is directly perpendicular to target, there's only one road
+    if cell[1] >= cols - 1 or cell[0] >= rows - 1:
+        return 1
+    # otherwise, there's two  roads to take, the answer is the sum of both
+    memo[cell] = robot_move((cell[0] + 1, cell[1]), rows, cols, memo) + robot_move(
+        (cell[0], cell[1] + 1), rows, cols, memo
+    )
+    # we save the answer in the memo for future referance and return it
+    return memo[cell]
```
Comment (Safi222) [inline solutions/robot.py:47]: Clear Logic

File: solutions/robot.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        >>> robot(3, 7)
+        28
+        >>> robot(4, 3)
+        10
+        >>> robot(23, 12)
+        193536720
+    """
+    # assert the values of rows and clos and make sure they are ints and
+    # greater than 1
+    assert isinstance(rows, int) and rows > 0, "Rows variables has to \
```
Comment (Safi222) [inline solutions/robot.py]: The assertion is clear, but here's a suggestion to use  ValueError instead of AssertionError to handle argument validation, it's a clearer signal that the provided argument value is incorrect.

**Review by Safi222 — APPROVED** [review]
The code is well-structured, and the logic is easy to follow , The tests are comprehensive.
Great job!

File: solutions/robot.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+at the top-left corner (i.e., grid[0][0]). The robot tries to move
+to the bottom-right corner (i.e., grid[m - 1][n - 1]). The robot
+can only move either down or right at any point in time.
+
+Given the two integers m and n, return the number of possible
+unique paths that the robot can take to reach the bottom-right
+corner.
+
+Constraints: The test cases are generated so that the answer will
+be less than or equal to 2 * 109.
```
Comment (RafaaAli) [inline solutions/robot.py:14]: The code should include the author's name, and the creation date!

**Comment by Safi222** [comment]
Well-structured! 
clear logic!
